### DISCUSSION TITLE
PowerSet X for Devices without Power Monitoring

### DISCUSSION THREAD

**Discussion by danielr1996** [arendst/Tasmota#10640] [Ideas]
**Have you looked for this feature in other issues and in the docs?**  
I've searched the issues and found #3208 which suggests that the feature is currently not available.
The [Documention for Power Calibration](https://tasmota.github.io/docs/Power-Monitoring-Calibration/) has no info on the supported devices.

**Is your feature request related to a problem? Please describe.**  
I cannot validate if this is already possible because i don't have the required devices and want to check the features before buying them. 

Power Monitoring with devices like the Sonoff Pow R2 is nice for devices with _varying_ loads such as computers, washing machines, tvs, etc. but quite unneccesary for devices with _static_ loads like light bulbs because the sonoff pow would always display roughly the same value or zero if turned off.

**Describe the solution you'd like**  
**Devices with Power Measurement**
Devices that already measure the power consumption would not be affected at all.

**Devices with on/off switch**
On Devices with an on/off switch but no power measurement i want to set the power consumption manually (e.g. by measuring with kill-a-watt, another sonoff pow, reading the spec for a light bulb,...) with `PowerSet X`.
Now if the device is turned on i'd like the device to report a power usage of X, if the device is turned off a power usage of 0.

**Devices with a dimmer**
~~In addition to the on/off behaviour (on meaning 100% dimmer position) the device would report X * dimmer position as power consumption (Assuming that dimmer position correlates linearly to the power usage).~~
As stated in the comments a reliable value for every dimmer setting can not be calculated for dimmers.

**Describe alternatives you've considered**  
Alternativly one could just monitor the on/off state and use that in custom software or ioBroker to calculate the power consumption afterwards. Currently i'm using a custom tasmota build with the prometheus extension, a prometheus instance that scrapes the sonoff pow and a grafana dashboard that displays the scraped values. Calculating the power usage afterwards would require either
1) setting up another software that modifies the scraped values (this might be easier if already using iobroker but im currently not planning on doing so.
2) using the [Transformations Feature](https://grafana.com/docs/grafana/latest/panels/transformations/) in Grafana and calculating the value inside grafana, but this has the drawback that the desired values are only correct in grafana but not other software that might use prometheus as a datasource.

**Additional context**  
I know this is rather an edge case but i hope there are other people that are facing the same problem.

I'm not experienced with programming of embedded software but i would get my hands dirty and try to and submit a pull request implementing this feature. 

**(Please, remember to close the issue when the problem has been addressed)**

**Comment by ascillato2** [comment]
This request was originally made by @danielr1996 on https://github.com/arendst/Tasmota/issues/9314

Moved to Github Discussion for further exploration of this idea, but this new feature of github didn't copy all the previous comment. Sorry about that. Bug already reported to Github.

**Comment by Frogmore42** [comment]
The other challenge is that currently the energy/power sensor code can only handle a single device, while the standard relay control can handle up to 8.  I have some devices that are just on on/off controls.  I have a Grafana transformation/calculation that calculates on time, which I find to be a more interesting number than energy usage.  If all you want reported is xx W when the switch is on that would not require extra storage, but I am not sure it would provide a whole lot of value either.

My take on it is: if I care about the power usage, I need to measure it and I am willing to pay the $10-15 extra per circuit to do that.  I am doing the same thing with a low voltage circuit by using an INA219 to measure power and energy.  It is not that I really care about the energy use from a cost perspective (or even an environmental one), but I do care about what the energy usage tells me about the health of the mechanical system it is powering and that it really is doing so.

**Comment by barbudor** [comment]
In such case, as the power would not be effectively measured by the devce, I would handle energy at the home application level (in my case node-red+influx)

**Reply by ortegafernando** [reply]
> In such case, as the power would not be effectively measured by the devce, I would handle energy at the home application level (in my case node-red+influx)

Yes, it is better by doing it in node-red influxdb.

**Comment by schrej-zz** [comment]
Hi,
I still think, that this makes sense and from a first glance it should be relatively easily to implement into the current code.
Regarding "buying a power monitoring device", you forget, that you (usually) cannot add other sensons like temperature sensors on these devices, as the GPIOs might have high voltage.
I think, a SETOPTION and using POWERSET and maybe also VOLTAGESET are sufficent to calculate and feed the other values.
Existing function can be reused. One just need to set POWER (and VOLTAGE) instead of measureing them.
Using Node Red or HA, or whatever overlying system might be a solution, but I would prefer so get these virtual values right from the device.

**Reply by arendst** [reply]
I'll build a dummy energy meter just for joy. Hold on.

**Comment by arendst** [comment]
Try latest development version.

Enable feature by selecting any GPIO as ``Option A2``. Then use default values or enter your own values using commands ``VoltageSet``, ``CurrentSet``, ``PowerSet`` and/or ``FrequencySet`` keeping in mind that P = U * I.

See file ``xnrg_20_dummy.ino`` for additonal information.

![image](https://user-images.githubusercontent.com/11044339/113579482-abc90880-9624-11eb-9ac9-344161e249cc.png)

![image](https://user-images.githubusercontent.com/11044339/113579408-8f2cd080-9624-11eb-8f1c-490541ab0576.png)

Do not forget to enable ``SetOption21 1``

**Reply by chicato83** [reply]
How to set volt for every relay?

**Reply by ascillato** [reply]
Voltage is the same for both.

**Reply by chicato83** [reply]
Oh now I got it. Thanks.

When I add it to has with MQTT I got zero info from relays.

**Reply by sibero80** [reply]
> Try latest development version.
> 
> Enable feature by selecting any GPIO as `Option A2`. Then use default values or enter your own values using commands `VoltageSet`, `CurrentSet`, `PowerSet` and/or `FrequencySet` keeping in mind that P = U * I.
> 
> See file `xnrg_20_dummy.ino` for additonal information.
> 
> ![image](https://user-images.githubusercontent.com/11044339/113579482-abc90880-9624-11eb-9ac9-344161e249cc.png)
> 
> ![image](https://user-images.githubusercontent.com/11044339/113579408-8f2cd080-9624-11eb-8f1c-490541ab0576.png)
> 
> Do not forget to enable `SetOption21 1`

Hello all and happy new year.  I'm trying to implement this with some of my relays for devices with known power draw. However my electric grid does not provide a good quality constant voltage (South America) so I was wondering if there is a way to link the voltage reading to another Tasmota Device that can do power metering?  I'm using Home Assistant, so I was wondering the following approach, but not sure how to achieve this:
1. set `VoltageSet `to `none` on the relay device.
2. on the power metering device, set a rule or command that "duplicates" the voltage reading to a MQTT message that overrides de relay's empty voltage.

Any thoughts on how to achieve this?

**Comment by schrej-zz** [comment]
Wow! This was quite fast!
I just tested it with a Sonoff Basic and I think, it still has a little error.
The wattage (and Wh) is still calculated, no matter, if the relay ist on or off.
Should only have wattage, when relay is ON.

I don't dare to ask, if this could also be extended to more relays in one device.

Joerg

**Reply by schrej-zz** [reply]
Just to be sure, I now also tested with a WEMOS D1 Mini and the GENERIC template.
Result ist the same. Wattage still shown and measured no matter if relays are ON or OFF.

I tried to figure out the issue in the code. I think, it's the ENERGY.power_on flag in xnrg_20_dummy.ino.
If I'm right, setting SO21 results in having ENERGY.power_on allways being on, no matter, if a relay is ON or OFF.
(See xdrv_03_energy.ino)
Maybe another or an additional variable is needed in xnrg_20_dummy.ino to figure out the relay state.
Joerg

**Comment by arendst** [comment]
Fixed in latest dev version. Don't ask for more!

**Reply by schrej-zz** [reply]
Hi @arendst,
thanks. Looked in the code and it's more than I expected as you added measurement for up to three relays ("missused" phases if I got it right). I just expected a test just on flag TasmotaGlobal.power being 0 or greater.
I will wait for the next build run and will give feedback.
Thanks again for the great and fast support (as allways)
Joerg

**Comment by schrej-zz** [comment]
Hi,
tested now with a Gitpod compiled version.
All three relays are working and show wattage etc. when POWER ON and 0 when POWER OFF.
Combined energy values are also working as expected.
All relays can only being set to the same wattage, etc.  though.
But this is due to the fact that usually Power-/Current-/Voltage-/Set is used for calibration of a whole device and not for each separate relay.
But I don't care and if I or someone else wants more, we should support by our own pull request.
Thanks again for the great support and this phantastic software.
Joerg

**Comment by KhurramFHassan** [comment]
I cant find this in the pre-built 9.4.0 release. Can it be added to a custom build?

**Reply by barbudor** [reply]
@KhurramFHassan 
As per [BUILDS.md](https://github.com/arendst/Tasmota/blob/master/BUILDS.md) it should be there
Did you configured a GPIO as `OptionA 2` ?

**Comment by schrej-zz** [comment]
Hi,
it's in the standard build. Don't forget to assign a GPIO pin to "Option A2".
Joerg

**Comment by KhurramFHassan** [comment]
Hmm, I have already assigned gpio12 to Option A2 but I get nothing on the main page. I am actually compiling my own firmware (with slave/client support). I wonder if there is some conflict with this.
![main](https://user-images.githubusercontent.com/45386894/117985091-d7d95700-b351-11eb-9d1c-34ec0893abeb.jpg)
![moduleconfig](https://user-images.githubusercontent.com/45386894/117985138-e0319200-b351-11eb-897c-7cae3306ca21.jpg)

**Reply by barbudor** [reply]
If you do your own compile, be sure to have USE_ENERGY_DUMMY

**Reply by sfromis** [reply]
You can also verify if you have USE_ENERGY_DUMMY included. Normally, decode-status should provide this info, but the tool has not yes been expanded for 8 sets of feature bits. However, the output of `Status 4` includes the bit in the Features array:
`"Features":["00000809","8FDAC787","04368001","000000CF","010013C0","C000F981","00004004","00001000","00000020"]`
If the last entry, as here, is 00000020 (or has that bit set), USE_ENERGY_DUMMY is included.

**Reply by KhurramFHassan** [reply]
Status 4 gives
`{"ProgramSize":602,"Free":400,"Heap":21,"ProgramFlashSize":4096,"FlashSize":4096,"FlashChipId":"164020","FlashFrequency":40,"FlashMode":3,"Features":["00000809","8FDAC787","04368001","000000CF","010013C0","C001F981","00004004","00001000","00000020"],"Drivers":"1,2,3,4,5,6,7,8,9,10,12,16,18,19,20,21,22,24,26,27,29,30,31,35,37,45","Sensors":"1,2,3,4,5,6"}}`
So it is included but I don't get any power readings in the main page.

**Reply by sfromis** [reply]
I'm not seeing any relay pin in your config. You can't get fake "energy monitoring" without something to monitor.

**Reply by KhurramFHassan** [reply]
Ah, that was missing. I added a (non-existent) relay on D6 and Option A2 on D7 and the power data shows up. I would like to use VoltageSet and CurrentSet from a rule triggered by the data returned by the TasmotaClient. This data is:
`01:28:36.753 MQT: tele/homeupsstabilizer/SENSOR = {"Time":"2021-05-14T01:28:36","ENERGY":{"TotalStartTime":"2021-05-14T00:51:12","Total":0.000,"Yesterday":0.000,"Today":0.000,"Period":0,"Power":0,"ApparentPower":0,"ReactivePower":0,"Factor":0.00,"Voltage":0,"Current":0.000},"TasmotaClient":{"Voltage":222,"Current":1861,"RPow":326,"APow":413,"PF":0.78,"Freq":50,"IEner":413,"AEner":0}}`

My rule looks like:
`Rule1 ON TasmotaClient#Voltage DO VoltageSet %value% ENDON ON TasmotaClient#Current DO CurrentSet %value% ENDON`

USE_RULES is defined in user_config_override.h and I have turned rule1 on. Rule1 never gets executed; I have increased logging to 4 for weblog but I see nothing in the console which shows that the rule has executed and neither does it show any errors in the console. What am I doing wrong?

![image](https://user-images.githubusercontent.com/45386894/118184375-b2814180-b454-11eb-8225-d0cd78e36b71.png)

**Reply by sfromis** [reply]
Not sure why that does not trigger. As a bit of a longshot, I'd try:
`Rule1 ON tele-TasmotaClient#Voltage DO VoltageSet %value% ENDON ON tele-TasmotaClient#Current DO CurrentSet %value% ENDON`

**Reply by KhurramFHassan** [reply]
I used StatusSNS#TasmotaClient#Voltage etc as the trigger and its working. Now there is another problem which I have posted below.

**Comment by KhurramFHassan** [comment]
After a few hiccups, I am able to used OptionA2 on my "TasmotaClient-ized" device (nodemcu v3) thanks to sfromis and others above. Now I am trying to get data into the dummy power device from the tasmotaclient (which is an arduino running the emonlib library to measure voltage, current, power etc). The data returned by the tasmota client is as follows:
`{"StatusSNS":{.......,"TasmotaClient":{"Voltage":234,"Current":1.961,"RPow":306,"APow":459,"PF":0.66,"Freq":50,"IEner":459,"AEner":0}}`

The rule am using is
`Rule1 ON StatusSNS#TasmotaClient#Voltage DO VoltageSet %value% ENDON ON StatusSNS#TasmotaClient#RPow DO PowerSet %value% ENDON ON StatusSNS#TasmotaClient#Current DO CurrentSet %value% ENDON`

But only the voltage is changing in the main window; current and power remain at 0 even though the console windows shows that it is excecuting all 3 commands. Also, what are the commands to set apparent power and power factor? Thanks.
![image](https://user-images.githubusercontent.com/45386894/118267205-1c413000-b4d5-11eb-9e27-e8ff893e5b92.png)

**Reply by KhurramFHassan** [reply]
No replies, sigh! I have reverted to using nodered to update voltage, current and power values to domoticz :(

**Reply by barbudor** [reply]
I get the expected result with the values from your exemple

Please do not use screenshots for the logs but copy paste the text using `insert code` formatting

```
5:58:41.024 CMD: VoltageSet 231
15:58:41.026 SRC: WebConsole from 192.168.168.5
15:58:41.030 CMD: Group 0, Index 1, Command "VOLTAGESET", Data "231"
15:58:41.037 MQT: stat/nodemcu/VOLTAGESET = {"VoltageSetCal":23100}
15:58:52.872 CMD: CurrentSet 2306
15:58:52.876 SRC: WebConsole from 192.168.168.5
15:58:52.879 CMD: Group 0, Index 1, Command "CURRENTSET", Data "2306"
15:58:52.886 MQT: stat/nodemcu/CURRENTSET = {"CurrentSetCal":230600}
15:59:30.844 CMD: PowerSet 488
15:59:30.847 SRC: WebConsole from 192.168.168.5
15:59:30.850 CMD: Group 0, Index 1, Command "POWERSET", Data "488"
15:59:30.856 MQT: stat/nodemcu/POWERSET = {"PowerSetCal":48800}
```

![image](https://user-images.githubusercontent.com/2996042/118363882-aef5d380-b596-11eb-9027-2493f6d8642c.png)

**Reply by barbudor** [reply]
From your screenshot of the webgui I can read "OFF"
When power is OFF, no power measurement are supposed to be performed

**Reply by KhurramFHassan** [reply]
You are absolutely right. I turned the relay on and it started updating current and power as well. I though Setoption21 1 was supposed to take care of that. No worries, I will just keep it on all the time.

Thanks.

**Reply by meingraham** [reply]
P=IV thus when the "relay" is off, there is no current thus no power. Even though there is no actual relay in your use case, that's the way the software is written - no current, no power measurement.

There is no means to set the power factor that I am aware of. You can "calibrate" your power "readings" using https://tasmota.github.io/docs/Power-Monitoring-Calibration/#calibration-procedure

**Reply by barbudor** [reply]
@KhurramFHassan `SetOption21` is *only* about updating the voltage in the webgui when power is OFF

Indeed power factor is computed automatically
The full power formula is P = I x V x cos(phi) 
So when you provide the tripplet P, I, V the reactive factor is automatically computed and apparent/reactive power updated as needed.
If you enter `backlog voltageset 230; currentset 1000; powerset 230` you get a power factor of 1.00

**Reply by KhurramFHassan** [reply]
Thanks for the explanation :)

**Comment by ghost** [comment]
This feature is great. There is one thing missing for me though - the ability to set the power/load per relay.

My use case is as follows: I have a 2 gang wall light switch with no built in means for power monitoring. Connected to both of those relays inside the switch is a different sized bank of lights, hard-wired. One is around 450 watts, and another is around 250 watts.

The tasmota web UI as is, appears to allow for  a different power, voltage and current setting per relay. But I have tried many variations of things like "PowerSet1 450" "PowerSet 1 450" and so on, and cannot get different values to "stick" to each relay.

Is it at all possible to get this feature?

Thanks

**Reply by sfromis** [reply]
No, the calibration settings are not per channel. Also if using a real energy monitoring device there would be one calibration process for the device as such, before starting to use it to gather data.

**Reply by ghost** [reply]
That is a shame. Thanks for the quick answer though.
